add command for resetting agent state

aalemayhu opened the pull request:
Please see individual commits for details. This should be ready now.

aalemayhu committed:
cilium: add command for resetting agent state

Requires root privileges.

	$ cilium help cleanup
	Reset the agent state

	Usage:
	  cilium cleanup [flags]

	Flags:
	  -f, --force   Skip confirmation

	Global Flags:
	      --config string   config file (default is $HOME/.cilium.yaml)
	  -D, --debug           Enable debug messages
	  -H, --host string     URI to server-side API

Closes: #979 (Provide a cilium-agent flag to clean all state data)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
tests: use daemon cleanup flag in CIDR limit test

Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

raybejjani commented on daemon/daemon_cleanup.go:
```diff
@@ -0,0 +1,115 @@
+	fmt.Printf("- library code\n")
+	fmt.Printf("- endpoint state\n")
+	fmt.Printf("- CNI configuration\n")
+	fmt.Printf("The command is non-revertable, do you want to contine [y/n]?")
```
"Destructive" is typoed.

If you prefer, you can build this as a single string:
```golang
fmt.Printf("Desctructive operation! This will remove\n" +
	"- all BPF maps\n" +
	"- mounted bpffs\n" +
	"- routes created by Cilium\n" +
	"- library code\n" +
	"- endpoint state\n" +
	"- CNI configuration\n" +
	"The command is non-revertable, do you want to contine [y/n]?")
```

raybejjani commented on daemon/daemon_cleanup.go:
```diff
@@ -0,0 +1,115 @@
+	out, err := exec.Command("umount", bpf.GetMapRoot()).CombinedOutput()
+	if err == nil {//TODO: make quite on no BPF filesystem present
+		fmt.Printf("umounted bpf map root %s\n", out)
+	}
```
`out` might be multiline (I imagine this is more likely in the error case) and that might mess up the Printf by being multiline by accident.

aalemayhu committed:
pkg/bpf: add function for unmounting the BPF filesystem

Will be used in follow up to reset daemon state.

Also expose isBpffs().

Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aanm commented on daemon/daemon_cleanup.go:
```diff
@@ -0,0 +1,178 @@
+func runCleanup() {
+	// Abort if the agent is running
+	if _, err := os.Stat(defaults.SockPath); !os.IsNotExist(err) {
+		fmt.Fprintf(os.Stderr, "Agent should not be running when cleaning up\n"+
```
Is the socket always removed when the agent crashes for example? What about creating a `.pid` file?

aalemayhu replied:
Using pidfile, thanks.

aanm commented on daemon/daemon_cleanup.go:
```diff
@@ -0,0 +1,178 @@
+
+func cleanupConfirmed() bool {
+	fmt.Printf("Destructive operation! This will remove:\n" +
+		"- all BPF maps\n" +
```
It will be helpful to inform the user where the maps are located, as well the directories where the library code, endpoint statue, cni configuration.

Give him a more detailed information on which exact things are going to be modified. (this includes the routes and links that are going to be removed on the system)

Ideally you'll need to check everything before, presented to the user if they agree what cilium will delete and then delete the exact same information that you've shown to the user.

aalemayhu replied:
So I changed the output to include more of the paths. Are you suggesting to list everything that is going to be removed?

aanm replied:
yes, regarding the paths, only the parent paths

aanm commented on daemon/daemon_cleanup.go:
```diff
@@ -0,0 +1,178 @@
+func removeDirs() error {
+	dirs := []string{defaults.RuntimePath, defaults.LibraryPath}
+	for _, dir := range dirs {
+		if err := os.RemoveAll(dir); err != nil {
```
to be precocious throw an error if `dir` is `/` before doing a `RemoveAll`

aalemayhu replied:
Fixed

aanm commented on daemon/daemon_cleanup.go:
```diff
@@ -0,0 +1,178 @@
+		if !strings.Contains(name, ciliumPrefix) && name != tunnelMap {
+			continue
+		}
+		path := fmt.Sprintf("%s/%s", mapDir, name)
```
see https://golang.org/pkg/path/filepath/#Join

aalemayhu replied:
Fixed

aanm commented on daemon/main.go:
```diff
@@ -354,6 +355,7 @@ func init() {
 		"version", false, "Print version information")
 	flags.Bool(
 		"pprof", false, "Enable serving the pprof debugging API")
+	flags.BoolVar(&cleanAll, "clean-all", false, "Reset Cilium state")
```
I don't get this, before you're stopping the program if the socket was running, but you're adding the flag in the daemon.

I think you would like to add this option to cilium CLI, the daemon should not be running.

aalemayhu replied:
Fixed, but named it `cleanup` for now. If you have a better name please share.

aalemayhu committed:
daemon: add a pidfile to determine if the agent is running

Using the socket is not reliable for several reasons. i) checking for
presence of socket while agent is starting. ii) agent dies but socket
remains. There are probably other scenarios but experienced those ones.

This can also be used in the future to prevent multiple agents from
running on the same host. See #2061 (daemon: fail to start if there is
already a running agent).

Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

tgraf commented on cilium/cmd/cleanup.go:
```diff
@@ -0,0 +1,233 @@
+	"github.com/spf13/cobra"
+	"github.com/vishvananda/netlink"
+	"io/ioutil"
+	"path/filepath"
```
Can you split the imports into the usual 3 sections?

golang std

cilium internal

external

aalemayhu replied:
Fixed

tgraf commented on cilium/cmd/cleanup.go:
```diff
@@ -0,0 +1,233 @@
+}
+
+func showWhatWillBeRemoved(routes map[int]netlink.Route, links map[int]netlink.Link) {
+	fmt.Printf("Destructive operation! This will remove:\n"+
```
```
Warning: Destructive operation. You are about to remove:
```

aalemayhu replied:
Fixed

tgraf commented on cilium/cmd/cleanup.go:
```diff
@@ -0,0 +1,233 @@
+var force bool
+
+const (
+	ciliumPrefix = "cilium"
```
Please rename to `ciliumLinkPrefix` and use `cilium_` instead of `cilium`

aalemayhu replied:
Fixed

tgraf commented on cilium/cmd/cleanup.go:
```diff
@@ -0,0 +1,233 @@
+	if links, err := netlink.LinkList(); err == nil {
+		for _, link := range links {
+			linkName := link.Attrs().Name
+			if !strings.Contains(linkName, ciliumPrefix) {
```
We want to match on `lxcXXXX` interfaces here as well

aalemayhu replied:
Fixed

raybejjani commented on cilium/cmd/cleanup.go:
```diff
@@ -0,0 +1,242 @@
+
+func runCleanup() {
+	// Abort if the agent is running
+	if _, err := os.Stat(defaults.PidFilePath); !os.IsNotExist(err) {
```
This is surprisingly subtle ;) I was midway through explaining how it wouldn't work before realising that it does. A comment pointing out that the err == nil successful `Stat` case is handled correctly might be useful.
Separately, however, we don't print errors if they are real but not IsNotExist. This might happen if permissions are wrong for the CLI to see that file (maybe?). The error message would be confusing in that case :(

aalemayhu replied:
Can add to the comment, but not sure I understand your second part. You are supposed to run the command with root privilege, why would wrong permissions matter here?

raybejjani replied:
Oh, you're right. We check that earlier so this won't ever happen. Totally fine then!

raybejjani commented on cilium/cmd/cleanup.go:
```diff
@@ -0,0 +1,242 @@
+	for _, dir := range dirs {
+		if err := os.RemoveAll(dir); err != nil {
+			// In the unlikely case one of the constants is the root directory, abort.
+			if dir == "/" {
```
Shouldn't this check occur before the `RemoveAll` above?

aalemayhu replied:
ah missed that one. Thanks.

raybejjani commented on cilium/cmd/cleanup.go:
```diff
@@ -0,0 +1,242 @@
+		if !confirmCleanup() {
+			return
+		}
+	}
```
Does this order of ifs mean we won't show what's changing with --force? I would expect force to only suppress the confirmation, and the command to always show the changes. What was the reasoning here?

aalemayhu replied:
I wanted the force option to be quiet but will change it.

raybejjani replied:
Ah, I thought so. I can see the argument to keep it quiet. shrug. I'm not too attached to my proposal.

raybejjani commented on daemon/main.go:
```diff
@@ -127,6 +132,22 @@ func main() {
 	}
 }
 
+func setupPidFile() {
```
Do we need to handle another agent still running? this would be a pidfile still in place that points to a running PID. We don't seem to handle another agent instance's socket already existing in initEnv() so maybe this isn't a situation we handle anyway.
At first glance, I would expect us to do a check similar to the one for `cilium cleanup` above and refuse to start the agent if another is already running.
@aanm do you know?

aanm replied:
Agreed

aalemayhu replied:
Can this be done separately?

aalemayhu replied:
Created a new issue for this https://github.com/cilium/cilium/issues/2061

raybejjani reviewed (changes requested):
Nothing too big, just an open question about agent behaviour on startup with another agent running.

aanm commented on daemon/main.go:
```diff
@@ -127,6 +132,22 @@ func main() {
+	}
+
+	// Handle the cleanup
+	sig := make(chan os.Signal, 1)
```
I think you should move this to a new function called clean up. We might use it to do more cleanups besides `os.Remove(defaults.PidFilePath)`

aalemayhu replied:
fixed

raybejjani commented on daemon/main.go:
```diff
@@ -127,6 +132,22 @@ func main() {
+
+	// Handle the cleanup
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, syscall.SIGQUIT, syscall.SIGINT, syscall.SIGHUP, syscall.SIGTERM)
```
Oh, also, I think this `Notify` means the default handlers for these signals are no longer auto-exit. We'd need to handle them explicitly here, otherwise we can never stop the agent :/ I also wouldn't expect SIGHUP to be an exit signal unless it's left with the default handler (in my experience, it's used to reload configs but it's supposed to be a quit, I think).

aalemayhu replied:
IIRC systemd is sending SIGHUP on stop or restart. Can you elaborate on what problem you are seeing? I am not following you.

raybejjani replied:
The SIGHUP thing is fine then, assuming we're always running under systemd.
The actual problem is that this is the only place we call `signal.Notify` and I think that suppresses any default handling of signals. The default would be to quit, I think. The goroutine we spawn here will read the channel, and keep removing the pidfile, but never quit. This means we wouldn't actually exit when systemd sent us a QUIT or whatever.
I tried this with
```golang
package main

import (
        "fmt"
        "os"
        "os/signal"
        "syscall"
        "time"
)

func main() {
        sig := make(chan os.Signal, 1)
        signal.Notify(sig, syscall.SIGQUIT, syscall.SIGINT, syscall.SIGHUP, syscall.SIGTERM)
        go func() {
                for s := range sig {
                        fmt.Printf("saw %+v\n", s)
                }
        }()
        time.Sleep(30 * time.Second)
}
```

aalemayhu replied:
@raybejjani Thanks for explaining. So is it enough to `os.Exit(0)` here or do you meaning something else by explicit?

raybejjani replied:
Basically, yes. It would be good to also log that we are exiting (and maybe which signal we received). In my dummy test it seems like the default handlers say something related to the signal received, and the exit code also seems different:
```bash
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ sigtest
hi!
Terminated
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ echo $?
143
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ sigtest
hi!
Hangup
vagrant@cilium-k8s-master:~/go/src/github.com/cilium/cilium$ echo $?
129
```
Longer term, we will likely want to do something different. The current behaviour, and the behaviour with this PR will both suppress `defer` statements, and generally not do cleanup tasks. The suggestion to have a cleanup function above would be a first step in having a more robust system later.

aalemayhu replied:
Would be great if you could create an issue for it. Thanks.

raybejjani replied:
yessir! https://github.com/cilium/cilium/issues/2063

raybejjani commented on cilium/cmd/cleanup.go:
```diff
@@ -0,0 +1,242 @@
+var force bool
+
+const (
+	ciliumLinkPrefix = "cilium_"
```
This shouldn't block this PR but are these prefixes/names defined already in the relevant packages, and would it be more robust to export them from there and use them here?

aalemayhu replied:
Will add to my todo list for replacing {init,join_ep}.sh

tgraf replied:
Yes, good idea to clean this up as we go go-native with these interactions.

raybejjani commented on cilium/cmd/cleanup.go:
```diff
@@ -0,0 +1,242 @@
+
+	showWhatWillBeRemoved(routes, links)
+	if !force {
+		if !confirmCleanup() {
```
A nit, but this could be one line `if !force && !confirmCleanup()`

raybejjani commented on cilium/cmd/cleanup.go:
```diff
@@ -0,0 +1,242 @@
+	if strings.HasPrefix(linkName, ciliumLinkPrefix) {
+		return true
+	}
+	return strings.HasPrefix(linkName, hostLinkPrefix) && len(linkName) == hostLinkLen
```
purely a style thing, the return and if could be merged into:
```golang
strings.HasPrefix(linkName, ciliumLinkPrefix) || strings.HasPrefix(linkName, hostLinkPrefix) && len(linkName) == hostLinkLen
```

aalemayhu replied:
Will merge them, but what I don't like about it is if we add more conditions it just gets uglier.

raybejjani replied:
Hmm, I see, yes. It's up to you, definitely not a big deal. I also don't know how many more we'll add in the near term. Also, unlike my example above, we can have a newline to group things a bit
```golang
return strings.HasPrefix(linkName, ciliumLinkPrefix) ||
	strings.HasPrefix(linkName, hostLinkPrefix) && len(linkName) == hostLinkLen
```

tgraf replied:
This could be written beautifully if golang would support switch case fall throughs.

raybejjani replied:
@tgraf it does! https://play.golang.org/p/htXuXmh3Ws
https://golang.org/ref/spec#Fallthrough_statements

How would that apply here, though? We could have an OR by using a case with comma separated clauses but I don't see how it would be more compact.

raybejjani commented on cilium/cmd/cleanup.go:
```diff
@@ -0,0 +1,242 @@
+
+func findRoutesAndLinks() (map[int]netlink.Route, map[int]netlink.Link, error) {
+	routesToRemove := map[int]netlink.Route{}
+	linksToRemove := map[int]netlink.Link{}
```
I don't think we use any of the indexes (the int, here) anywhere. If you preferred to treat these maps as a set you can use `map[*netlink.Route]bool` and `map[*netlink.Link]bool` (or also non-pointer versions, I think). You would then iterate over them in `removeRoutesAndLinks` with only `for route := range routes` etc.

aalemayhu replied:
Fine as is, thanks.

tgraf replied:
I'm fine with how it's currently done, matter of personal preference.

raybejjani commented on cilium/cmd/cleanup.go:
```diff
@@ -0,0 +1,242 @@
+	if len(routes) > 0 {
+		fmt.Printf("- routes\n")
+		for _, v := range routes {
+			fmt.Printf("%v\n", v)
```
Is there any chance we can make this printout a little more human readable? The route object was ok but the links one was difficult to understand because it's full of nils. Unfortunately, there doesn't seem to be a nice String() function on any of these objects... I tried "%+v" and it printed the field names, at least, which make it easier to read.
```
- routes
{Ifindex: 187 Dst: 169.254.254.0/24 Src: <nil> Gw: <nil> Flags: [] Table: 254}
- links
&{LinkAttrs:{Index:11 MTU:1450 TxQLen:0 Name:lxc15cdb HardwareAddr:5e:8d:e4:9d:64:02 Flags:up|broadcast|multicast RawFlags:69699 ParentIndex:10 MasterIndex:0 Namespace:<nil> Alias: Statistics:0xc4215066d0 Promisc:0 Xdp:<nil> EncapType:ether Protinfo:<nil> OperState:up} PeerName:}
&{LinkAttrs:{Index:183 MTU:1450 TxQLen:0 Name:lxc41b3f HardwareAddr:ea:57:7f:12:87:05 Flags:up|broadcast|multicast RawFlags:69699 ParentIndex:182 MasterIndex:0 Namespace:<nil> Alias: Statistics:0xc421506bc8 Promisc:0 Xdp:<nil> EncapType:ether Protinfo:<nil> OperState:up} PeerName:}
&{LinkAttrs:{Index:186 MTU:1500 TxQLen:1000 Name:cilium_net HardwareAddr:56:20:e9:9c:b5:5b Flags:up|broadcast|multicast RawFlags:69827 ParentIndex:187 MasterIndex:0 Namespace:<nil> Alias: Statistics:0xc4215070ec Promisc:0 Xdp:<nil> EncapType:ether Protinfo:<nil> OperState:up} PeerName:}
&{LinkAttrs:{Index:187 MTU:1500 TxQLen:1000 Name:cilium_host HardwareAddr:7a:0a:65:7e:d5:e9 Flags:up|broadcast|multicast RawFlags:69827 ParentIndex:186 MasterIndex:0 Namespace:<nil> Alias: Statistics:0xc4215075dc Promisc:0 Xdp:<nil> EncapType:ether Protinfo:<nil> OperState:up} PeerName:}
```

aalemayhu replied:
Let's leave this as is and if it gets too much we can pretty print later.

tgraf replied:
Please file a microtask GH issue to write nice output functions. It's a great entry level task.

aalemayhu replied:
Issue created https://github.com/cilium/cilium/issues/2104

raybejjani commented on daemon/main.go:
```diff
@@ -127,6 +132,27 @@ func main() {
+	cleanUp()
+}
+
+func cleanUp() {
```
I would rename this to something like initCleanup and also call it in the main function, right after `setupPidFile`. This is mostly because it has longer lived side-effects (the signal handling goroutine) that are a bit more global.

aalemayhu replied:
Looks better this way.

raybejjani commented on daemon/main.go:
```diff
@@ -127,6 +132,27 @@ func main() {
+	signal.Notify(sig, syscall.SIGQUIT, syscall.SIGINT, syscall.SIGHUP, syscall.SIGTERM)
+	go func() {
+		for s := range sig {
+			log.Infof("exiting received signal %v", s)
```
You could structure this with `log.WithField("signal", s).Info("Exiting due to signal")`. Also, I think our convention is to capitalise the first letter of log lines, if you fancy being consistent with that.

aalemayhu replied:
Is there some Go tooling that can detect or enforce this? Would be nice to be warned before / after compiling.

raybejjani replied:
Not easily, no. For the capitalisation thing we could get way with a regex but for the function usage the best I can think of is having a rule for `eg` and see if it changes the code as part of CI (the way we do with `go fmt`).
`eg` https://rakyll.org/eg/

tgraf replied:
Could probably be added to golint though.

raybejjani reviewed (approved):
Small comments that you can safely ignore :)
